### DISCUSSION TITLE
Invert deterministic operations from measured variables

### DISCUSSION THREAD

**Discussion by ricardoV94** [aesara-devs/aeppl#119] [Ideas]
**Epistemic status:** I am not sure about the usefulness of this, other than for variable transforms which we already have a solution for. So this may very well fit in the cute but perhaps useless category.

***

When Aeppl is asked to derive a logprob graph from a deterministic 1-1 operation based on some upstream random/measurable variable, we have to create a graph that reverses these deterministic operations (adding extra logprob and jacobian terms if needed): 

```python
import aesara.tensor as at
from aeppl import joint_logprob

x = at.normal()
y = at.log(x)
z = at.normal(y)

y_vv = y.clone()
z_vv = z.clone()
joint_logprob({y: y_vv, z: z_vv})
```

This specific case is not implemented yet, but would be after #26

Other logprob terms like that of `z` are then allowed to depend on `y`. **However**, they are not allowed to depend `x` (which in the logprob term of `z` should be replaced by `exp(y_vv)`) or in any other intermediate variables between `y` and `x` if there were more. For instance the following is not possible:

```python
x = at.normal()
y = at.log(x)
z = at.normal(x)

y_vv = y.clone()
z_vv = z.clone()
joint_logprob({y: y_vv, z: z_vv})
```

That type of logp graph can currently be achieved via the TransformValuesOpt but that feels like a clunky add-on. If the above were allowed, we would have a very natural way of defining variable transformations in Aeppl. For more complex transformations we could still use specialized Op's that specify what the inverse graph should be:

```python
x = at.normal()
y = ComplicatedTransformOp()(x)
z = at.normal(x)
```

And the transformation would still be part of the generative graph, without the need for the extra rewrite.

I wonder how feasible this would be, and whether it opens the door for more interesting features or just falls in the "cute" category.

**Comment by brandonwillard** [comment]
The expectations for this library are that both examples **should** be supported, so there's no need to question the relevance of the idea.

Side note:

Regarding these specific examples, the easiest way to support `y = at.exp(x)` is to include a rewrite that simply matches such a graph and replaces it with a log-normal `RandomVariable`.   You could manually add such a rewrite for the example above and we would expect it to work.

This non-generalized-transform-based approach maintains a simplicity that generic transforms can't.  More specifically, it's much easier to build further rewrites that operate on or expect log-normals, because it's much easier to identify a log-normal when it's just a `RandomVariable`.  The same goes for basic convolutions (e.g. normal-normal).

Ultimately, we want to support such rewrites at least as much as the more generic transform approach.  It should be possible for people  to work with simple graphs that&mdash;as directly as is reasonably possible&mdash;map to a high-level representation of a model.

**Comment by ricardoV94** [comment]
Just to be clear the problem is not with replacing `y = at.exp(at.random.normal(name="x")) -> at.random.lognormal(name="y")`, but to allow other measured variables to depend on the original `x` variable, which in the logprob graph would mean depending on `at.log(y_value)`

This involves acting on the value variables directly (at least that's what we are doing so far for TransformedRVs), but we may not know which value variables are ultimately associated with intermediate IR variables

**Reply by brandonwillard** [reply]
This seems to touch up on https://github.com/aesara-devs/aeppl/pull/78, which introduces a graph-based means of "acting upon" value variables.

In that scenario, one could return a `dict` in a local rewrite and affect arbitrary changes in value variables quite easily.

Aside from that, I'm not entirely clear on the dependencies you're implying here.

In an example like
```python
x = at.normal()
y = at.exp(x)
z = at.normal(y)
```
with values on/in terms of `y` and `z` (i.e. the joint distribution of `y` and `z`), `y` resolves to a log-normal density through a simple rewrite, and `z` to a normal conditioned on the value provided for `y`.

Although, `z` depends on `y` in the underlying probabilistic model, there is no relevant dependency between the two aside from the information from `y` necessary to derive the conditional distribution of `z`.

If we were trying to rewrite **the model graph itself** (i.e. to produce a new, equivalent model graph in terms of a log-normal `y`), then we would need to do a lot more, and it sounds like you might be interested in this, unless I'm missing something.

**Comment by rlouf** [comment]
I think this should be turned into an issue. To re-state the issue, if I take your second example:

```python
import aesara
import aesara.tensor as at
from aeppl.joint_logprob import joint_logprob

srng = at.random.RandomStream(0)
x_rv = srng.normal()
y_rv = at.log(x_rv)
z_rv = srng.normal(x_rv)

y_vv = y_rv.clone()
z_vv = z_rv.clone()
logprob = joint_logprob({y_rv: y_vv, z_rv: z_vv})
```

AePPL warns that there are still random variables in the graph:

```python
aesara.dprint(logprob)
        
# Sum{acc_dtype=float64} [id A]
#  |MakeVector{dtype='float64'} [id B]
#    |Sum{acc_dtype=float64} [id C]
#    | |Elemwise{add,no_inplace} [id D]
#    |   |Check{sigma > 0} [id E]
#    |   | |Elemwise{sub,no_inplace} [id F]
#    |   | | |Elemwise{sub,no_inplace} [id G]
#    |   | | | |Elemwise{mul,no_inplace} [id H]
#    |   | | | | |TensorConstant{-0.5} [id I]
#    |   | | | | |Elemwise{pow,no_inplace} [id J]
#    |   | | | |   |Elemwise{true_div,no_inplace} [id K]
#    |   | | | |   | |Elemwise{sub,no_inplace} [id L]
#    |   | | | |   | | |Elemwise{exp,no_inplace} [id M]
#    |   | | | |   | | | |<TensorType(float64, ())> [id N]
#    |   | | | |   | | |TensorConstant{0.0} [id O]
#    |   | | | |   | |TensorConstant{1.0} [id P]
#    |   | | | |   |TensorConstant{2} [id Q]
#    |   | | | |Elemwise{log,no_inplace} [id R]
#    |   | | |   |Elemwise{sqrt,no_inplace} [id S]
#    |   | | |     |TensorConstant{6.283185307179586} [id T]
#    |   | | |Elemwise{log,no_inplace} [id U]
#    |   | |   |TensorConstant{1.0} [id P]
#    |   | |All [id V]
#    |   |   |Elemwise{gt,no_inplace} [id W]
#    |   |     |TensorConstant{1.0} [id P]
#    |   |     |TensorConstant{0.0} [id X]
#    |   |<TensorType(float64, ())> [id N]
#    |Sum{acc_dtype=float64} [id Y]
#      |Check{sigma > 0} [id Z]
#        |Elemwise{sub,no_inplace} [id BA]
#        | |Elemwise{sub,no_inplace} [id BB]
#        | | |Elemwise{mul,no_inplace} [id BC]
#        | | | |TensorConstant{-0.5} [id BD]
#        | | | |Elemwise{pow,no_inplace} [id BE]
#        | | |   |Elemwise{true_div,no_inplace} [id BF]
#        | | |   | |Elemwise{sub,no_inplace} [id BG]
#        | | |   | | |<TensorType(float64, ())> [id BH]
#        | | |   | | |normal_rv{0, (0, 0), floatX, False}.1 [id BI]
#        | | |   | |   |RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7F4CD01BFA00>) [id BJ]
#        | | |   | |   |TensorConstant{[]} [id BK]
#        | | |   | |   |TensorConstant{11} [id BL]
#        | | |   | |   |TensorConstant{0.0} [id O]
#        | | |   | |   |TensorConstant{1.0} [id P]
#        | | |   | |TensorConstant{1.0} [id P]
#        | | |   |TensorConstant{2} [id BM]
#        | | |Elemwise{log,no_inplace} [id BN]
#        | |   |Elemwise{sqrt,no_inplace} [id BO]
#        | |     |TensorConstant{6.283185307179586} [id BP]
#        | |Elemwise{log,no_inplace} [id BQ]
#        |   |TensorConstant{1.0} [id P]
#        |All [id BR]
#          |Elemwise{gt,no_inplace} [id BS]
#            |TensorConstant{1.0} [id P]
#            |TensorConstant{0.0} [id BT]
```

**Comment by rlouf** [comment]
I am not completely sure about the design of `AeMCMC`&rsquo;s inteface yet, but I think that what Ricardo suggests in the first comment would be useful to us. Indeed, in the [current design iteration](<https://github.com/aesara-devs/aemcmc/pull/68#issuecomment-1266479262>) `AeMCMC` will return a stream of samplers, and callers should be able to inspect the graph that was used to sample as:

```python
import aesara.tensor as at
from aemcmc.basic import construct_sampler

srng = at.random.RandomStream(0)
Y_rv = my_model(srng)
y_vv = Y_rv.clone()

sampling_steps, updates, initial_values = construct_sampler(srng, {Y_rv: y_vv})
print(sampling_steps.model)
# Prints the model graph for instance
```

The NUTS sampler currently operates on the original graph, and AePPL handles the transforms at the value variable level, so the transformation is not immediately obvious to callers, unless they inspect the `transform` attribute of the `SamplingStep` instance.

Let&rsquo;s give an example of logprob with transforms on a simple model:

```python
import aesara.tensor as at
from aemcmc.basic import construct_sampler

import aeppl
from aeppl.transforms import LogTransform, TransformValuesRewrites


srng = at.random.RandomStream(0)
mu_at = at.scalar('mu')
sigma_rv = at.random.lognormal(1.)
Y_rv = at.random.normal(mu_at, sigma_rv)

sigma_vv = sigma_rv.clone()
y_vv = Y_rv.clone()

rewrites = TransformValuesRewrites({sigma_vv: LogTransform()})
logprob = aeppl.joint_logprob({Y_rv: y_vv, sigma_rv: sigma_vv}, extra_rewrites=rewrites)
```

When I would like instead to be able to transform the original graph to the following and condition on `sigma_rv_tr`:

```python
srng = at.random.RandomStream(0)
mu_at = at.scalar('mu')
sigma_rv = at.random.lognormal(1.)
sigma_rv_tr = at.log(sigma_rv)
Y_rv = at.random.normal(mu_at, sigma_rv)

sigma_vv_tr = sigma_rv_tr.clone()
y_vv = Y_rv.clone()

logprob = aeppl.joint_logprob({Y_rv: y_vv, sigma_rv_tr: sigma_vv_tr})
sigma_vv = at.exp(sigma_vv_tr)
```

So NUTS can operate on this space, and then returns `at.exp(sigma_rv_tr)` to the caller. AeMCMC can then return this kind of information:

```python
print(sampling_steps[sigma_rv].model)
# Model sampled
#
# mu_at = at.scalar('mu')
# sigma_rv = at.random.lognormal(1.)
# sigma_rv_tr = at.log(sigma_rv)  <-- conditionned on
# Y_rv = at.random.normal(mu_at, sigma_rv)  <--- conditionned on
#
# Returns
#
# sigma_vv = at.exp(sigma_vv_trans)
```

More fundamentally, I don&rsquo;t see any reason why the bijections needed by algorithms in the HMC family should have a special status among graph rewrites. In the end, these transforms should be implemented as a function that returns the transformed graph, the value variables to condition on and the value variables in the original space. It is fine to me if `AePPL`'s interface to transforms stay the same, but I would like to be able to do the kinds of conditioning @ricardoV94 mentions in his original post.

**Reply by ricardoV94** [reply]
You can also introduce a "LogTransformedNormal” in the graph that will behave like that (I think). It's just a bit less elegant I think

**Reply by rlouf** [reply]
Is there a clear path to implementing what you were suggesting in the original post (example 2) or are there things we'd need to figure out?

**Reply by ricardoV94** [reply]
> Is there a clear path to implementing what you were suggesting in the original post (example 2) or are there things we'd need to figure out?

I haven't figured out what the code to achieve that might look like.

**Reply by brandonwillard** [reply]
As I said above, https://github.com/aesara-devs/aeppl/pull/78 is a first step in these directions.  In other words, we will probably need to track variable/value pairs **in-graph**.
